src/gromo/utils/utils.py

```diff
@@ -170,6 +170,137 @@ def line_search(cost_fn: Callable, verbose: bool = True) -> tuple[float, float]:
     return factor, min_loss
 
 
+def mini_batch_gradient_descent(
+    model: nn.Module | Callable,
+    cost_fn: Callable,
+    X: torch.Tensor,
+    Y: torch.Tensor,
+    lrate: float,
+    max_epochs: int,
+    batch_size: int,
+    parameters: Iterable | None = None,
+    fast: bool = False,
+    eval_fn: Callable | None = None,
+    verbose: bool = True,
+) -> tuple[list[float], list[float]]:
+    """Mini-batch gradient descent implementation
+    Uses AdamW with no weight decay and shuffled DataLoader
+
+    Parameters
+    ----------
+    model : nn.Module
+        pytorch model or forwards function
+    cost_fn : Callable
+        cost function
+    X : torch.Tensor
+        input features
+    Y : torch.Tensor
+        true labels
+    lrate : float
+        learning rate
+    max_epochs : int
+        maximum epochs
+    batch_size : int
+        batch size
+    parameters: iterable | None, optional
+        list of torch parameters in case the model is just a forward function, by default None
+    fast : bool, optional
+        fast implementation without evaluation, by default False
+    eval_fn : Callable | None, optional
+        evaluation function, by default None
+    verbose : bool, optional
+        print info, by default True
+
+    Returns
+    -------
+    tuple[list[float], list[float]]
+        train loss history, train accuracy history
+    """
+    loss_history, acc_history = [], []
+    full_loss = []
+    gradients = []
+
+    dataset = torch.utils.data.TensorDataset(X, Y)
+    dataloader = torch.utils.data.DataLoader(dataset, batch_size=batch_size, shuffle=True)
+
+    if not isinstance(model, nn.Module):
+        if (parameters is None) or (len(parameters) < 1):
+            raise AttributeError(
+                "When the model is just a forward function, the parameters argument must not be None or empty"
+            )
+    else:
+        parameters = model.parameters()
+        saved_parameters = list(model.parameters())
+    optimizer = torch.optim.AdamW(parameters, lr=lrate, weight_decay=0)
+
+    for epoch in range(max_epochs):
+        correct, total, epoch_loss = 0, 0, 0
+        for x_batch, y_batch in dataloader:
+            optimizer.zero_grad(set_to_none=True)
```

Comment:
If I understand it correctly you added the `set_to_none` option when you were searching for the bug fixed with safe_forward. This is probably not useful anymore ? If it is unnecessary removing it will porbably make the function faster as it avoid re-affecting memory each time.